api/src/api/users.rs

```diff
@@ -0,0 +1,145 @@
+        .await;
+    match existing_user {
+        Ok(Some(_)) => {
+            return HttpResponse::BadRequest().body("user with this username already exists");
+        }
+        Ok(None) => {}
+        Err(e) => {
+            return HttpResponse::InternalServerError().body(format!(
+                "failed to query database for existing user. error: {}",
+                e
+            ));
+        }
+    }
+    let existing_user = user_collection
+        .find_one(doc! { "email": &body.email })
+        .await;
+    match existing_user {
+        Ok(Some(_)) => {
+            return HttpResponse::BadRequest().body("user with this email already exists");
+        }
+        Ok(None) => {}
+        Err(e) => {
+            return HttpResponse::InternalServerError().body(format!(
+                "failed to query database for existing user. error: {}",
+                e
+            ));
+        }
+    }
+
+    // create a new user document
+    // first, hash password
+    // TODO: permissions?
+    let hashed_password =
+        bcrypt::hash(&body.password, bcrypt::DEFAULT_COST).expect("failed to hash password");
+    let user_bson = doc! {
+        "username": &body.username,
+        "email": &body.email,
+        "password": hashed_password,
+    };
+
+    // save new user to database
+    match user_collection.insert_one(user_bson).await {
+        Ok(_) => {
+            return HttpResponse::Ok().body("successfully created new user");
+        }
+        Err(e) => {
+            return HttpResponse::InternalServerError()
+                .body(format!("failed to insert user into database. error: {}", e));
+        }
+    }
+}
+
+#[derive(serde::Serialize)]
+struct User {
+    username: String,
+    email: String,
+    id: String,
+}
+
+#[get("/users")]
+pub async fn get_users(db: web::Data<Database>) -> HttpResponse {
+    let user_collection: Collection<mongodb::bson::Document> = db.collection("users");
+    let users = user_collection.find(doc! {}).await;
+
+    match users {
+        Ok(mut cursor) => {
+            let mut user_list = Vec::new();
+            while let Some(user) = cursor.next().await {
+                match user {
+                    Ok(doc) => {
+                        // extract only what we want from the user document
+                        let user = User {
+                            username: doc
+                                .get_str("username")
+                                .map(|val| val.to_string())
+                                .unwrap_or_default(),
+                            email: doc
+                                .get_str("email")
+                                .map(|val| val.to_string())
+                                .unwrap_or_default(),
+                            id: doc
+                                .get_object_id("_id")
+                                .map(|id| id.to_string())
+                                .unwrap_or_default(),
```

Comment:
I wonder if `unwrap_or_default` is the right choice. How severe/critical/interesting is it to encounter an Err from these `.get_*` methods?

Reply:
It would be a problem if there were ever a user in the database without a username, email, or id, so it would at least be good if the caller saw a descriptive error and could go clean things up.

Reply:
Ok so I guess the question is, does that warrant a 4xx or 5xx response to the caller or should it be logged as a warning for the operators to see? I don't know what the right answer is.